### ISSUE TITLE
[FG:InPlacePodVerticalScaling] Scheduling race condition

### ISSUE THREAD

**Issue by tallclair** [kubernetes/kubernetes#126891]
What is the desired behavior when a pod is resized in the middle of scheduling? The current behavior is that after a certain point the scheduler would not pick up the updated resources and proceed to schedule a pod to a node, even if it no longer fits. The kubelet will then reject the pod in admission (`OutOfCPU` or `OutOfMemory`).

I think we have 3 options:

1. Proceed with the current approach, and accept (document) this as a possible failure mode
2. Forbid resizing un-scheduled pods (use case is questionable)
3. Minimize (but don't eliminate) the risk by checking for changes before binding. There will always be a race condition here, since the bind request is not transactional.

I'm leaning towards the first option.

/sig node
/priority important-longterm

**Comment by SergeyKanzhelev** [comment]
/kind feature

Is there an option to allow to resize pods which failed admission and make kubelet to try to readmit it?

**Comment by tallclair** [comment]
> Is there an option to allow to resize pods which failed admission and make kubelet to try to readmit it?

No, but if we had a mechanism to send rejected pods back to the scheduler (like we discussed the other day) then that could cover this case.

**Comment by tallclair** [comment]
Thinking about this more, I think we should just embrace option 1 and document this case, but not invest in preventing or minimizing it. I definitely don't think it should block beta.

**Comment by ffromani** [comment]
> > Is there an option to allow to resize pods which failed admission and make kubelet to try to readmit it?
> 
> No, but if we had a mechanism to send rejected pods back to the scheduler (like we discussed the other day) then that could cover this case.

I had a sketch of a design document not so long ago about kubelet-driven pod rescheduling which fizzled out for lack of bandwidth, but it seems it's time to update it and highlight the VPA use case. I'll give it some time in a couple weeks time.

**Comment by esotsal** [comment]
> Thinking about this more, I think we should just embrace option 1 and document this case, but not invest in preventing or minimizing it. I definitely don't think it should block beta.

+1 with option 1

**Comment by iholder101** [comment]
> 2. Forbid resizing un-scheduled pods (use case is questionable)

For my education, do we have such a use-case in mind?

**Comment by tallclair** [comment]
See the discussion on https://github.com/kubernetes/enhancements/pull/4704#discussion_r1768680167, and [accompanying proposal](https://docs.google.com/document/d/1O9rDeYhGWMbtg-DQmhqKvNsQSHWJSwvSEeNpXTKMvZg/edit).

We agreed this is out of scope for InPlacePodVerticalScaling beta.

**Comment by kerthcet** [comment]
> Forbid resizing un-scheduled pods (use case is questionable)

Some inputs here, maybe not 100% related, but we do hope to resize the pod during scheduling. So the general situtation looks like: when serving AI workloads, A model could be served with H100, A100 or even L40s, however, the static scheduling model couldn't fit this, so we have another CRD to prioritize the accelerators as a slice, and we use node label for node selection. Then in scheduling, we hope to dynamic change the resource requests to meet our needs, like
```
- pick A100 first, for both cost and performance consideration
- then pick H100, a bit more expensive, but if we have enough H100s, why not
- last pick L40s, we have no choice because high-end accelerators are stockout
```

Nowadays, we could change the cpu/memory with VPA enabled, but only these two resources IIRC. 
/sig scheduling

**Comment by kerthcet** [comment]
cc @ahg-g for suggestions as working on serving as well.
